src/main/java/games/strategy/net/nio/Encoder.java

```diff
@@ -90,4 +91,17 @@ private void write(final MessageHeader header, final ObjectOutputStream out, fin
     }
     out.reset();
   }
+
+  /**
+   * Provide access to the raw buffer.
+   */
+  private static final class ByteArrayOutputStream2 extends ByteArrayOutputStream {
```

Comment:
Good question.  As the Javadoc states, the only purpose of this class is to expose the `buf` field to the enclosing class.  I can't really think of a better name that says, "we're subverting visibility to get direct access to the underlying buffer." :smile:  I'm not even sure why a copy of the buffer couldn't be passed to `SocketWriteData` (via `ByteArrayOutputStream#toByteArray()`) unless it's for some performance benefit.  Of course, there could be some hidden requirement that the original buffer is able to be modified by whoever uses `SocketWriteData`...?

Looking at how this type is used in `Encoder`, it's really just an implementation-specific output stream because all the call sites where it's used simply require an `OutputStream`.  If this class were still a top-level type, I would suggest calling it `EncoderOutputStream`.  But since it's now nested, that would be redundant, so what about just `OutputStream` (which would be rendered as the type `Encoder$OutputStream`)?  This file currently does not import `java.io.OutputStream`, so there would be no aliasing, but it could still lead to confusion for the casual code reader.